src/aggregators/rdirect.jl

```diff
@@ -0,0 +1,138 @@
+  execute_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  generate_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  register_next_jump_time!(integrator, p, integrator.t)
+  nothing
+end
+
+# setting up a new simulation
+function (p::RDirectJumpAggregation)(dj, u, t, integrator) # initialize
+  initialize!(p, integrator, u, integrator.p, t)
+  register_next_jump_time!(integrator, p, t)
+  nothing
+end
+
+############################# Required Functions #############################
+
+# creating the JumpAggregation structure (tuple-based constant jumps)
+function aggregate(aggregator::RDirect, u, p, t, end_time, constant_jumps,
+                    ma_jumps, save_positions, rng; kwargs...)
+
+    # handle constant jumps using function wrappers
+    rates, affects! = get_jump_info_fwrappers(u, p, t, constant_jumps)
+
+    build_jump_aggregation(RDirectJumpAggregation, u, p, t, end_time, ma_jumps,
+                          rates, affects!, save_positions, rng; num_specs = length(u), kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::RDirectJumpAggregation, integrator, u, params, t)
+    fill_rates_and_sum!(p, u, params, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+"execute one jump, changing the system state and updating rates"
+function execute_jumps!(p::RDirectJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u)
+
+    # update rates
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+"calculate the next jump / jump time"
+function generate_jumps!(p::RDirectJumpAggregation, integrator, u, params, t)
+    # if no more events possible there is nothing to do
+    sum_rate = p.sum_rate
+    if nomorejumps!(p, sum_rate)
+        return nothing
+    end
+    @unpack rng, cur_rates, max_rate = p
+
+    num_rxs = length(cur_rates)
+    rx = trunc(Integer, rand(rng) * num_rxs)+1
+    while cur_rates[rx] < rand(rng) * max_rate
+        # global num_rejections += 1
+        rx = trunc(Integer, rand(rng) * num_rxs)+1
+    end
+    # global num_reactions += 1
+
+    p.next_jump = rx
+
+    # update time to next jump
+    p.next_jump_time = t + randexp(p.rng) / sum_rate
+    # if (p.next_jump_time >= p.end_time)
+    #     println("Average number of rejections per reaction: $(num_rejections/num_reactions)")
+    # end
+    nothing
+end
+
+
+######################## SSA specific helper routines #########################
+function update_dependent_rates!(p::RDirectJumpAggregation, u, params, t)
+    @inbounds dep_rxs = p.dep_gr[p.next_jump]
+    @unpack ma_jumps, rates, cur_rates, sum_rate = p
+    @inbounds for rx in dep_rxs
+        sum_rate -= cur_rates[rx]
+        @inbounds new_rate = calculate_jump_rate(ma_jumps, rates, u,params,t,rx)
+        sum_rate += new_rate
+        cur_rates[rx] = new_rate
+        if new_rate > p.max_rate
+            p.max_rate = new_rate
```

Comment:
Hmm, so the upper bound is never recalculated from scratch, and always just grows? I’d imagine it could get pretty bad then, like if you have 100000 molecules in the spatial example all started in the first voxel.

Reply:
Maybe you should keep a counter, and periodically regenerate the max rate if it hasn’t increased for a certain number of calls?